### PR TITLE
Add verification request viewer in devtools

### PR DESCRIPTION
**Pull request by foldleft**
It's per-room, and only for room-based events rather than device-based ones.
Screenshot 2020-02-21 at 17 16 36

### CODE REVIEW

**Comment by foldleft** [comment]
… I'm not sure it's really properly firing the change event and updating. Can anyone see a bug in my logic?

File: src/components/views/dialogs/DevtoolsDialog.js [hunk 1]
```diff
@@ -605,12 +614,97 @@ class ServersInRoomList extends React.PureComponent {
...
+        request.on("change", updateState);
+        return () => request.off("change", updateState);
+    }, [request]);
+
+    /* Keep re-rendering if there's a timeout */
+    useEffect(() => {
+        if (request.timeout == 0) return;
+
+        const id = setInterval(() => {
+           setTimeout(request.timeout);
```
Comment (t3chguy) [inline src/components/views/dialogs/DevtoolsDialog.js]: what is this meant to be doing?

Comment (foldleft) [reply src/components/views/dialogs/DevtoolsDialog.js]: keeps the timeout up to date

Comment (foldleft) [reply src/components/views/dialogs/DevtoolsDialog.js]: to be more clear, `request.timeout` is defined as such:

```
    /** The current remaining amount of ms before the request should be automatically cancelled */
    get timeout() {
        const requestEvent = this._getEventByEither(REQUEST_TYPE);
        if (requestEvent) {
            const elapsed = Date.now() - this.channel.getTimestamp(requestEvent);
            return Math.max(0, VERIFICATION_REQUEST_TIMEOUT - elapsed);
        }
        return 0;
    }
```

ie it's actually running code

Comment (t3chguy) [reply src/components/views/dialogs/DevtoolsDialog.js]: oh right its just using the default of 0ms, `setImmediate` might be clearer here

Comment (foldleft) [reply src/components/views/dialogs/DevtoolsDialog.js]: ah, `setTimeout` here is a state function, not `window.setTimeout`. Maybe I should clarify that

Comment (t3chguy) [reply src/components/views/dialogs/DevtoolsDialog.js]: my bad, completely missed that

File: src/components/views/dialogs/DevtoolsDialog.js [hunk 2]
```diff
@@ -605,12 +614,98 @@ class ServersInRoomList extends React.PureComponent {
...
+
+function VerificationRequest({txnId, request}) {
+    const [, updateState] = useState();
+    const [timeout, setRequestTimeout] = useState(request.timeout);
+
+    /* Re-render if something changes state */
+    useEffect(() => {
+        request.on("change", updateState);
+        return () => request.off("change", updateState);
+    }, [request]);
```
Comment (t3chguy) [inline src/components/views/dialogs/DevtoolsDialog.js]: can you use `useEventEmitter(request, "change", updateState);` as Encryption Panel does

File: src/components/views/dialogs/DevtoolsDialog.js [hunk 3]
```diff
@@ -605,12 +615,97 @@ class ServersInRoomList extends React.PureComponent {
...
+    componentWillUnmount() {
+        const cli = this.context;
+        cli.off("crypto.verification.request", this.onNewRequest);
+    }
+
+    render() {
+        const cli = this.context;
+        const room = this.props.room;
+        const inRoomChannel = cli._crypto._inRoomVerificationRequests;
+        const inRoomRequests = (inRoomChannel._requestsByRoomId || new Map()).get(room.roomId) || new Map();
```
Comment (t3chguy) [inline src/components/views/dialogs/DevtoolsDialog.js]: this is a bit unfathomable

Comment (foldleft) [reply src/components/views/dialogs/DevtoolsDialog.js]: yeah. bring on the day we get nullable variables

**Pull request merged**